packages/jest-runtime/src/transform.js

```diff
@@ -110,93 +112,179 @@ const readCacheFile = (filePath: Path, cachePath: Path): ?string => {
+  } else {
+    let preprocessor;
+    if (!config.scriptPreprocessor) {
+      preprocessor = null;
     } else {
-      content = preprocessor.process(content, filename, config);
-      if (options && options.instrument) {
-        content = options.instrument(content);
+      // $FlowFixMe
+      preprocessor = require(config.scriptPreprocessor);
+      if (typeof preprocessor.process !== 'function') {
+        throw new TypeError(
+          'Jest: a preprocessor must export a `process` function.',
+        );
       }
-      content = wrap(content);
-      writeCacheFile(cachePath, content);
     }
-  } else if (options && options.instrument) {
-    content = wrap(options.instrument(content));
+    preprocessorCache.set(config, preprocessor);
+    return preprocessor;
+  }
+};
+
+const stripShebang = content => {
+  // If the file data starts with a shebang remove it. Leaves the empty line
+  // to keep stack trace line numbers correct.
+  if (content.startsWith('#!')) {
+    return content.replace(/^#!.*/, '');
   } else {
-    content = wrap(content);
+    return content;
   }
+};
 
-  const script = new vm.Script(content, {
-    displayErrors: true,
-    filename,
-  });
+const instrument = (content: string, filename: Path): string => {
+  // NOTE: Keeping these requires inside this function reduces a single run
+  // time by 2sec if not running in `--coverage` mode
+  const babel = require('babel-core');
+  const babelPluginIstanbul = require('babel-plugin-istanbul').default;
+
+  if (babel.util.canCompile(filename)) {
+    return babel.transform(content, {
+      filename,
+      plugins: [babelPluginIstanbul],
+      retainLines: true,
+      babelrc: false,
+    }).code;
+  }
+  return content;
+};
+
+const cachedTransformAndWrap = (
+  filename: Path,
+  config: Config,
+  content: string
+) => {
+  const preprocessor = getPreprocessor(config);
+  const cacheFilePath = getFileCachePath(filename, config, content);
+  // Ignore cache if `config.cache` is set (--no-cache)
+  let result = config.cache ? readCacheFile(filename, cacheFilePath) : null;
+
+  if (result) {
+    return result;
+  }
 
-  cache.set(key, script);
-  return script;
+  result = content;
+
+  if (preprocessor && shouldPreprocess(filename, config)) {
+    result = preprocessor.process(result, filename, config);
+  }
+  if (shouldInstrument(filename, config)) {
+    result = instrument(result, filename, config);
+  }
+
+  result = wrap(result);
+  writeCacheFile(cacheFilePath, result);
+  return result;
+};
+
+const transformAndBuildScript = (filename: Path, config: Config): vm.Script => {
+  const content = stripShebang(fs.readFileSync(filename, 'utf8'));
+  let wrappedResult;
+
+  if (
+    shouldPreprocess(filename, config) || shouldInstrument(filename, config)
+  ) {
+    wrappedResult = cachedTransformAndWrap(filename, config, content);
+  } else {
+    wrappedResult = wrap(content);
+  }
+
+  return new vm.Script(wrappedResult, {displayErrors: true, filename});
```

Comment:
TODO: better error handling around this call. All SyntaxError in the transformed source code will throw here